design/metadata-handling.md

```diff
@@ -0,0 +1,668 @@
+
+#### Metadata Specifications
+
+The `metaData` field contains a list of items that will render data in different
+ways. The following objects are available:
+
+- **key**: This is compulsory and is the key used for the rendered value in
+  the metadata map.
+- **string**: renders the given string as value in the metadata
+- **objectName** : renders the name of the object that matches the type given in
+  the `object` field.
+- **index**: renders the index of the current object, with the offset from the
+  `offset` field and using the step from the `step` field.
+- **ip**: renders an ip address based on the index, based on the `start` value
+  if given or using `subnet` to calculate the start value, and checking that
+  the rendered value is not over the `end` value. The increment is the `step`
+  value.
+- **bareMetalHostMACByName**: renders the MAC address of the BareMetalHost that
+  matches the name given as value.
+
+The **key** is required and one of the object is required. If there are multiple
+objects, the priority order will apply as listed above.
+
+#### networkData specifications
+
+The `networkData` field will contain three items :
+
+- **links**: a list of layer 2 interface
+- **networks**: a list of layer 3 networks
+- **services** : a list of services (DNS)
+
+##### Links specifications
+
+The object for the **links** section list can be one of:
+
+- **ethernet**: an ethernet interface
+- **bond**: a bond interface
+- **vlan**: a vlan interface
+
+The **links/ethernet** object contains the following:
+
+- **type**: Type of the ethernet interface
+- **id**: Interface name
+- **mtu**: Interface MTU
+- **macAddress**: an object to render the MAC Address
+
+The **links/ethernet/type** can be one of :
+
+- bridge
+- dvs
+- hw_veb
+- hyperv
+- ovs
+- tap
+- vhostuser
+- vif
+- phy
+
+The **links/ethernet/macAddress** object can be one of:
+
+- **string**: with the desired Mac given as a string
+- **bareMetalHostMACByName**: with the interface name from BareMetalHost
+  hardware details.
+
+The **links/bond** object contains the following:
+
+- **id**: Interface name
+- **mtu**: Interface MTU
+- **macAddress**: an object to render the MAC Address
+- **bondMode**: The bond mode
+- **bondLinks** : a list of links to use for the bond
+
+The **links/bond/bondMode** can be one of :
+
+- 802.1ad
+- balance-rr
+- active-backup
+- balance-xor
+- broadcast
+- balance-tlb
+- balance-alb
+
+The **links/vlan** object contains the following:
+
+- **id**: Interface name
+- **mtu**: Interface MTU
+- **macAddress**: an object to render the MAC Address
+- **vlanId**: The vlan ID
+- **vlanLink** : The link on which to create the vlan
+
+##### The networks specifications
+
+The object for the **networks** section list can be one of:
+
+- **ipv4**: an ipv4 static allocation
+- **ipv4Dhcp**: an ipv4 dhcp based allocation
+- **ipv6**: an ipv6 static allocation
+- **ipv6Dhcp**: an ipv6 dhcp based allocation
+- **ipv6Slaac**: an ipv6 slaac based allocation
```

Comment:
The convention is usually to use the same case for all letters in an abbreviation, so I think we would want `ipv6DHCP` and `ipv6SLAAC` but otherwise this makes sense.

Reply:
done, thank you